### PR TITLE
Feature/logpoints

### PR DESCRIPTION
**Pull request by changsi-an**
The code on feature/logpoints have been reviewed by Jay Wang and Mike Haufman. I just synced master, merged it to the feature branch and resolved the conflicts. Everything else should be the same.

### CODE REVIEW

File: src/diagnostics/LogPointsManager.ts [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            return false;
+        }
+
+        if (vscode.window.activeTextEditor.document.uri.toString() !== uri.toString()) {
+            throw new Error("Invalid operation: cannot operate on an inactive text editor.");
+        }
+
+        const line = vscode.window.activeTextEditor.selection.active.line;
+        // tslint:disable-next-line:no-suspicious-comment
+        // TODO, see if we can get the column number from the editor.
```
Comment (changsi-an) [inline src/diagnostics/LogPointsManager.ts]: Converted to https://github.com/Microsoft/vscode-azureappservice/issues/192

File: package.json [hunk 2]
```diff
@@ -393,6 +437,11 @@
                     "when": "explorerResourceIsFolder == true",
                     "group": "zzz_@appServiceDeployCommands"
                 }
+            ],
+            "editor/context": [
+                {
+                    "command": "diagnostics.LogPoints.Toggle"
```
Comment (EricJizbaMSFT) [inline package.json]: should this also have `"when": "config.appservice.logpoints == true`

File: src/extension.ts [hunk 3]
```diff
@@ -212,6 +238,19 @@ export function activate(context: vscode.ExtensionContext): void {
...
+            const wizard = new LogPointsSessionAttach(outputChannel, client, node.treeItem.site, node.subscription);
+            await wizard.run();
+        }
+    });
+
+    initAsyncCommand(context, 'diagnostics.LogPoints.Toggle', async (uri: vscode.Uri) => {
+        logPointsManager.toggleLogpoint(uri);
+    });
+
+    context.subscriptions.push(vscode.commands.registerCommand('diagnostics.LogPoints.OpenScript', openScript));
```
Comment (EricJizbaMSFT) [inline src/extension.ts]: Use initCommand for this

(that way its tracked in our telemetry and errors are displayed properly)

File: src/diagnostics/LogPointsManager.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            // switch to a document that was opened during the debug session before.
+            return;
+        }
+
+        debugSessionManager.recoverLogpoints(documentUri);
+    }
+
+    // tslint:disable-next-line:no-empty
+    private cleanup(): void {
+        // tslint:disable-next-line:no-suspicious-comment
```
Comment (changsi-an) [inline src/diagnostics/LogPointsManager.ts]: Converted to https://github.com/Microsoft/vscode-azureappservice/issues/193

File: src/diagnostics/logPointsClient.ts [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+abstract class LogPointsDebuggerClientBase {
+    protected abstract call<ResponseType>(siteName: string, affinityValue: string, publishCredential: WebSiteModels.User, command: string): Promise<CommandRunResult<ResponseType>>;
+
+    protected makeCallAndLogException<ResponseType>(siteName: string, affinityValue: string,
+        publishCredential: WebSiteModels.User, command: string): Promise<CommandRunResult<ResponseType>> {
+        return this.call<ResponseType>(siteName, affinityValue, publishCredential, command)
+            .catch<CommandRunResult<ResponseType>>((err) => {
+                // tslint:disable-next-line:no-suspicious-comment
+                // TODO: re-enable
```
Comment (changsi-an) [inline src/diagnostics/logPointsClient.ts]: Converted to https://github.com/Microsoft/vscode-azureappservice/issues/194

File: src/diagnostics/structs/CommandRunResult.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        return this.exitCode === 0 && this.json && !this.json.error;
+    }
+
+    public get json(): ResponseType {
+        if (this._json === undefined) {
+            try {
+                this._json = JSON.parse(this.output);
+            } catch (err) {
+                // tslint:disable-next-line:no-suspicious-comment
+                // TODO: re-enable.
```
Comment (changsi-an) [inline src/diagnostics/structs/CommandRunResult.ts]: Convert to https://github.com/Microsoft/vscode-azureappservice/issues/194

File: package.json (hunk 2 above)
Comment (changsi-an) [reply package.json]: Good catch!

File: src/extension.ts (hunk 3 above)
Comment (changsi-an) [reply src/extension.ts]: `initCommand()` only transfer 1 parameter to the cakkback. My `openScript` expects two parameters. Hence the trouble.

Filed https://github.com/Microsoft/vscode-azureappservice/issues/195. Will look into a fix when I got a chance.

**Pull request merged**